Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3562,10 +3565,27 @@ auto AirIRGenerator::addArrayGet(uint32_t typeIndex, ExpressionType arrayref, Ex
     // https://bugs.webkit.org/show_bug.cgi?id=245405
     emitCCall(&operationWasmArrayGet, getValue, instanceValue(), addConstant(Types::I32, typeIndex), arrayref, index);
 
-    switch (elementType.kind) {
+    switch (resultType.kind) {
     case TypeKind::I32:
         result = g32();
         append(Move32, getValue, result);
+        switch (arrayGetKind) {
+        case GCOpType::ArrayGet:
+        case GCOpType::ArrayGetU:
+            break;
+        case GCOpType::ArrayGetS: {
+            size_t elementSize = *elementType.as<PackedType>() == PackedType::I8 ? sizeof(uint8_t) : sizeof(uint16_t);
+            uint8_t bitShift = (sizeof(uint32_t) - elementSize) * 8;
+            auto tmpForShift = g32();
+            append(Move32, Arg::imm(bitShift), tmpForShift);
+            addShift(Types::I32, Lshift32, result, tmpForShift, result);
+            addShift(Types::I32, Rshift32, result, tmpForShift, result);
```

Comment:
There really isn't an imm form of Lshift32?

Reply:
At least not with the `addShift` helper AFAIK, which is abstracting over different architectures. It might be possible to generalize `addShift` a bit but it didn't seem worth doing right now.